### PR TITLE
Fix for windows traversal attack

### PR DESCRIPTION
**Pull request by BenWilson2**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/BenWilson2/mlflow/pull/10647?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10647/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10647
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Changed from posixpath.basename to os.path.basename to protect in Windows

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Provide a fix for a potential windows traversal attack when using the mlflow.data API to load files in an Windows environment.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [X] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [X] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/data/http_dataset_source.py [hunk 1]
```diff
@@ -70,8 +69,8 @@ def load(self, dst_path=None) -> str:
                     f"Invalid filename in Content-Disposition header: {basename}. "
                     "It must be a file name, not a path."
                 )
-        elif path is not None and len(posixpath.basename(path)) > 0:
-            basename = posixpath.basename(path)
+        elif path is not None and len(os.path.basename(path)) > 0:
```
Comment (harupy) [inline mlflow/data/http_dataset_source.py:72]: What does `len(os.path.basename(path)) > 0` check?

Comment (harupy) [reply mlflow/data/http_dataset_source.py:72]: Does this check `basename` is not an empty string?

Comment (BenWilson2) [reply mlflow/data/http_dataset_source.py:72]: I tried to not mess with the original implementation for this fix. 

It's definitely an odd way of handling assignment of the basename. I'll adjust the logic to simplify a bit (without touching too much) just to make this more readable and prevent repeated unnecessary checks on that path.

Comment (harupy) [inline mlflow/data/http_dataset_source.py:72]: /ai Can you guess what `len(os.path.basename(path)) > 0` is for?

File: mlflow/data/http_dataset_source.py [hunk 2]
```diff
@@ -41,6 +40,25 @@ def url(self):
     def _get_source_type() -> str:
         return "http"
 
+    def _extract_filename(self, response) -> str:
+        """
+        Extracts a filename from the Content-Disposition header or the URL's path.
+        """
+        content_disposition = response.headers.get("Content-Disposition")
+        if content_disposition:
```
Comment (harupy) [inline mlflow/data/http_dataset_source.py:48]:
```suggestion
        if content_disposition := response.headers.get("Content-Disposition"):
```

**Review by harupy — APPROVED** [review]
LGTM, let's ask the reporter to validate the fix.

File: mlflow/data/http_dataset_source.py [hunk 3]
```diff
@@ -41,6 +40,25 @@ def url(self):
...
+            filename_match = next(re.finditer(r"filename=(.+)", content_disposition), None)
+            if filename_match:
```
[the remark this answers is not in the document]
Comment (BenWilson2) [reply mlflow/data/http_dataset_source.py:50]: agreed! great suggestion

**Comment by BenWilson2** [comment]
Received confirmation from researcher that this PR fixes the security issue. Merging.

**Pull request merged**